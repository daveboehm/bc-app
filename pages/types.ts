// --------------
// Private types
// --------------

type ImageType = {
    height: number;
    img?: HTMLElement;
    title: string;
    url: string;
    width: number;
}

type UrlType = {
    value: string;
    // @TODO: confirm available options for link target attr
    target: "_blank" | "parent" | "self";
}

// --------------
// External types
// --------------

export type EntryType = {
    description?: string;
    id: number;
    image: ImageType;
    title: string;
    url: UrlType;
}

export type EntriesPropsType = {
    entryIds: (string | undefined)[];
    hydratedEntries: (EntryType | undefined)[];
}

export type PaginationProps = {
    currentPage: number;
    itemsPerPage?: number;
    handlePageChange: (_pageNumber: number) => void;
    totalItems: number;
}