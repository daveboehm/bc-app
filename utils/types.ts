// --------------
// Private types
// --------------

type ImageType = {
    height: number;
    img?: HTMLElement;
    title: string;
    url: string;
    width: number;
}

type UrlType = {
    value: string;
    target: "_blank" | "_parent" | "_self" | "_top";
}

// --------------
// External types
// --------------

export type EntryType = {
    description?: string;
    id: number;
    image: ImageType;
    title: string;
    url: UrlType;
}

export type EntriesPropsType = {
    entryIds: (string | undefined)[];
    hydratedEntries: (EntryType | undefined)[];
}

export type PaginationProps = {
    currentPage: number;
    itemsPerPage?: number;
    handlePageChange: (_pageNumber: number) => void;
    totalItems: number;
}