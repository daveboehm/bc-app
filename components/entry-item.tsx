import React from "react"
import type { EntryType } from "../utils/types"
import styles from '../styles/Entry.module.css'
import Image from "next/image";

export const EntryItem = (props: EntryType) => {
    const { id, image, title, url } = props;

    if (!url || !id || !title) return null;

    return (
        <li className={styles.entryItem} data-testid="entry-item">
            <a className={styles.entryItemLink} href={url.value} target={url.target} title={image.url}>
                <div className={styles.entryImage}>
                    <Image src={image.url} alt={image.title} width={image.width} height={image.height} />
                </div>
                <h2 className={styles.entryTitle}>{title}</h2>
            </a>
        </li>
    )
}

export default EntryItem