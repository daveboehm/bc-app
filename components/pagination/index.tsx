import React, { useState, useEffect } from "react"
import type { PaginationProps } from "../../utils/types"
import { ITEMS_PER_PAGE } from "../../utils/constants"
import styles from '../styles/Pagination.module.css'

export const Pagination = (props: PaginationProps) => {
    const {
        currentPage,
        handlePageChange,
        itemsPerPage = ITEMS_PER_PAGE,
        totalItems
    } = props

    const [pageButtons, setPageButtons] = useState(1)

    useEffect(() => {
        const calcPageButtons = Math.ceil(totalItems / itemsPerPage)
        setPageButtons(calcPageButtons)
    }, [totalItems, itemsPerPage])

    return (
        <div className={styles.paginationContainer}>
            <div className={styles.buttonsContainer}>

                <button
                    data-testid="pagination-prev-button"
                    disabled={currentPage <= 0}
                    type="button"
                    className={styles.pageButtonInactive}
                    onClick={() =>handlePageChange(currentPage - 1)}>Prev</button>
                
                {pageButtons > 1 && [...Array(pageButtons)].map((_undef, pageIndex) => (
                    <button
                        data-testid="pagination-button"
                        type="button"
                        key={pageIndex}
                        className={pageIndex === currentPage ? styles.pageButtonActive : styles.pageButtonInactive}
                        onClick={() => handlePageChange(pageIndex)}>
                        {pageIndex + 1}
                    </button>
                    ))}

                <button
                    data-testid="pagination-next-button"
                    disabled={currentPage >= pageButtons-1}
                    type="button"
                    className={styles.pageButtonInactive}
                    onClick={() => handlePageChange(currentPage + 1)}>Next</button>
                
            </div>
        </div>
    )
}

export default Pagination