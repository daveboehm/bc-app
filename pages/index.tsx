import type { NextPage } from 'next'
import Head from 'next/head'
import Link from 'next/link';
import styles from '../styles/Home.module.css'

const Home: NextPage = () => {
  return (
    <div className={styles.container}>
      <Head>
        <title>Dave Boehm - Next app</title>
        <meta name="description" content="Take home code assessment" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <main className={styles.main}>
        <p>
          Hi there! tl;dr: just skip this rambling and click "Go To App" below.
        </p>

        <p>This is the second time I've ever spun up a nextjs app, and the first time I actually made something more than Hello World.</p>
        <p>I enjoyed nextjs immensely and spent a bit longer than the 4 hours expected of me on this mini project. I got hung up in one
          spot in particular: <code>getStaticProps()</code>.... I had written probably 100 lines of code by the time I found this
          method in their docs. :facepalm emoji: No wonder my client side fetch call didn't work at that point. Shame on me for not
          being more thorough in learning nextjs first.</p>
        <p>I included Typescript in here, moreso for my own preference, and because I wanted to see how nextjs and ts played together. I struggled a little trying To
          decipher the <code>'next/foo'</code> types in their declaration files, but I didn't spend much time on them either.</p>
        <p>Also tossed in some React Testing Library. I'm a huge fan of RTL over enzyme/chai/mocha/etc. I did not write all the tests
          I felt I should have for this app, but I think the tests I did include here demonstrate my style and competency. 
        </p>
        <p>I considered implementing sass/less, but I was already pretty unfamiliar with CSS modules and didn't want to overcomplicate something
          as easy as styling for an app of this size. I'm sure nextjs has a command that'll just do it all for me. I did not check.
        </p>
        <p>Lastly, I was forced to write this all on my gaming PC because I've completely forgotten my macbook password, and cannot log in
          or use it in any way. Not having a `cmd` key is devastating during development. This is also why I had to use npm instead of yarn. Well,
          not "had to", but the paths on my PC are a bit screwy and I simply didn't want to troubleshoot it.</p>



        <div>
          <Link href="/entries">
            <a className={styles.buttonLink}>Go To App</a>
          </Link>
        </div>
        
      </main>
    </div>
  )
}

export default Home
