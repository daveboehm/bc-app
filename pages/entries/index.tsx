import React, { useState, useEffect } from "react"
import type { NextPage } from 'next'
import Link from 'next/link'
import { EntryItem } from '../../components/entry-item'
import { Pagination } from '../../components/pagination'
import type { EntriesPropsType } from "../../utils/types"
import styles from '../../styles/Entry.module.css'
import { ENTRIES_URL, ENTRY_URL, ITEMS_PER_PAGE } from "../../utils/constants"

const Entries: NextPage | any = (props: EntriesPropsType) => {
    const { entryIds, hydratedEntries = [] } = props;
    
    const [currentPageEntries, setCurrentPageEntries] = useState(hydratedEntries)
    const [pageNum, setPageNum] = useState(0);

    useEffect(() => {
        const startIdx = pageNum * ITEMS_PER_PAGE
        const endIdx = startIdx + ITEMS_PER_PAGE
        const viewables = hydratedEntries.slice(startIdx, endIdx)

        setCurrentPageEntries(viewables);
        
    }, [pageNum, hydratedEntries])


    return (
        <main className={styles.main}>

            <Link href="/">
                <span className={styles.backLink}>Back to Home</span>
            </Link>

            <hr className={styles.hr} />

            <h1 className={styles.pageHeader}>All Entries ({entryIds.length})</h1>

            {hydratedEntries.length &&
                <ul className={styles.entries}>
                    {currentPageEntries.map((entry: any, idx: number) => (
                        <EntryItem {...entry} key={entry.id ?? idx} />
                    ))}
                </ul>
            }

            {(hydratedEntries.length > ITEMS_PER_PAGE) && 
                <>
                    <hr className={styles.hr} />

                    <Pagination currentPage={pageNum} handlePageChange={setPageNum} totalItems={entryIds.length} />
                </>
            }
           
        </main>
    )
}

// --------------
// Data Fetchin
// --------------
export const handleError = (err: any) => {
    console.error(err)
    // probablySendToErrorLoggerService("ListingPage", err)
}

const DEFAULT_ALL_THE_THINGS = {};
export async function getStaticProps() {

    // Fetch entry IDs
    const entriesResponse = await fetch(ENTRIES_URL, DEFAULT_ALL_THE_THINGS)
    if (!entriesResponse.ok) throw new Error("Super bad error you should totally care about")
    const entryIds: string[] = await entriesResponse.json()

    // 
    const entriesPromises = entryIds.map(async (id) => {
        const maybeEntry = await fetch(`${ENTRY_URL}?id=${id}`)
        if (maybeEntry.ok) { return await maybeEntry.json() }
        return null
    })
    
    // Resolve each entryById, remove bad eggs
    const hydratedEntries = await Promise.all(entriesPromises)
        .then(allEntries => allEntries.filter(noNulls => noNulls !== null))
        .catch(handleError)
    
    return { props: { entryIds, hydratedEntries } }
}

export default Entries
