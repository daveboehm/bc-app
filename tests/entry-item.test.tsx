import React from 'react'
import { render, screen, RenderResult } from "@testing-library/react";
import { EntryItem } from '../pages/components/entry-item';
import { EntryType } from '../pages/types';
import { MOCK_ENTRY_PROPS } from './__fixtures__/mock-data';

describe("<EntryItem />", () => {
    

    describe("happy path", () => {
        let props: EntryType
        let view: RenderResult
        
        beforeEach(() => {
            props = MOCK_ENTRY_PROPS
            view = render(<EntryItem {...props} />)
        })
        
        it("should match snapshot", () => {
            expect(view.container).toMatchSnapshot()
        });

        it("should use the correct alt text and src url", () => {
            const theImage = screen.getByRole("img")
            expect(theImage).toBeVisible()
            expect(theImage).toHaveAttribute("alt", props.image.title)
            expect(theImage).toHaveAttribute("src", props.image.url)
        })

        it("should use the correct link url", () => { 
            const theLink = screen.getByTitle(props.image.url)
            expect(theLink).toHaveAttribute("href", props.url.value)
        })
    })

    describe("sad path", () => {
        let props: EntryType | any // 'any' Type needed here to test bad data
        
        beforeEach(() => {
            props = MOCK_ENTRY_PROPS 
        })

        it.each([ "url", "id", "title" ])("should NOT render if missing %s prop", (_prop) => {
            delete props[_prop]

            const { container } = render(<EntryItem {...props} />)

            expect(container).toBeEmptyDOMElement();
        })
    })

});

