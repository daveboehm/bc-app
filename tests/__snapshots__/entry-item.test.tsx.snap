// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EntryItem /> happy path should match snapshot 1`] = `
<div>
  <li
    class="entryItem"
    data-testid="entry-item"
  >
    <a
      class="entryItemLink"
      href="https://www.heatpresszone.com/"
      target="_blank"
      title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
    >
      <div
        class="entryImage"
      >
        <img
          alt="Heat Press Zone"
          src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        />
      </div>
      <h2
        class="entryTitle"
      >
        Heat Press Zone
      </h2>
    </a>
  </li>
</div>
`;
